src/compose/MentionWarnings.js

```diff
@@ -0,0 +1,140 @@
+/* @flow strict-local */
+
+import React, { PureComponent } from 'react';
+import { connect } from 'react-redux';
+
+import type { Auth, Stream, Dispatch, Narrow, UserOrBot, Subscription } from '../types';
+import { getActiveUsersById, getAuth } from '../selectors';
+import { isPrivateNarrow } from '../utils/narrow';
+import * as api from '../api';
+
+import AnimatedScaleComponent from '../animation/AnimatedScaleComponent';
+import MentionedUserNotSubscribed from '../message/MentionedUserNotSubscribed';
+
+type State = {|
+  unsubscribedMentions: UserOrBot[],
+|};
+
+type SelectorProps = {|
+  auth: Auth,
+  usersById: Map<number, UserOrBot>,
+|};
+
+type Props = $ReadOnly<{|
+  narrow: Narrow,
+  stream: Subscription | {| ...Stream, in_home_view: boolean |},
+
+  dispatch: Dispatch,
+  ...SelectorProps,
+|}>;
+
+class MentionWarnings extends PureComponent<Props, State> {
+  state = {
+    unsubscribedMentions: [],
+  };
+
+  /**
+   * Check whether the message text entered by the user contains
+   * an @-mention to a user unsubscribed to the current stream, and if
+   * so, shows a warning.
+   *
+   * This function is expected to be called by `ComposeBox` using a ref
+   * to this component.
+   *
+   * @param message The message text entered by the user in the compose box.
+   */
+  handleMentionSubscribedCheck = async (message: string) => {
+    const { usersById, narrow, auth, stream } = this.props;
+
+    if (isPrivateNarrow(narrow)) {
+      return;
+    }
+    const unformattedMessage = message.split('**')[1];
+
+    // We skip user groups, for which autocompletes are of the form
+    // `*<user_group_name>*`, and therefore, message.split('**')[1]
+    // is undefined.
+    if (unformattedMessage === undefined) {
+      return;
+    }
+    const [userFullName, userId] = unformattedMessage.split('|');
+    const unsubscribedMentions = this.state.unsubscribedMentions.slice();
+    let mentionedUser: UserOrBot;
+
+    // eslint-disable-next-line no-unused-vars
+    for (const [id, user] of usersById) {
+      if (userId !== undefined) {
+        if (user.user_id === userId) {
+          mentionedUser = user;
+          break;
+        }
+      } else if (user.full_name === userFullName) {
+        mentionedUser = user;
+        break;
+      }
+    }
+    if (!mentionedUser || unsubscribedMentions.includes(mentionedUser)) {
+      return;
+    }
+
+    if (
+      !(await api.getSubscriptionToStream(auth, mentionedUser.user_id, stream.stream_id))
```

Comment:
I am inclined towards silently ignoring the error in this case, because the error will only confuse the user - did the @-mention get sent, or did it not?

Reply:
> the error will only confuse the user - did the @-mention get sent, or did it not?

Well, that uncertainty is pretty appropriate! After all, the point of these warnings is that the @-mention won't reach the mentioned user if they're not subscribed.

We'd like people to have confidence in general about whether someone else is going to get notified about their message or not. In cases where really don't know, the way to maintain that confidence is to say we don't know. Otherwise, when we *do* know and don't show a warning, people won't know whether to believe it, or whether instead this is another of those times that we failed to show them a warning before the surprising thing happened.

I think a toast would be the appropriate form to report the issue. `showErrorAlert` pops up a modal dialog box, which is more interruptive than seems appropriate for this condition. Compare:
https://material.io/components/dialogs#usage
https://material.io/components/snackbars#usage
(where "snackbar" is the current Material term for something very similar to a toast.)

Here's another iteration on the text: "Couldn't load information about ${user.full_name}".

Reply:
Yeah, makes sense.